### ISSUE TITLE
How to close OpenGL render contexts?

### ISSUE THREAD

**Issue by mix3d** [Kitware/vtk-js#1157]
What is the correct approach to "stop" or delete a render context?

In my Vue application, I am opening and closing vtk render windows. They do not seem to get garbage collected, since after 16 windows are opened, I get the following message:
`WARNING: Too many active WebGL contexts. Oldest context will be lost.`
This sounds like a memory leak issue.

How do I tell a render window to shut down and close its context, so it can be garbage collected properly?

**Comment by jourdain** [comment]
What are you currently doing?

Ideally you want to unbind any listener you have set and delete the renderWindow.

**Comment by mix3d** [comment]
Using the GenericRenderWindow which already handles creating and setting up the renderWindow object. There are no immediate functions that seem related to closing what it sets up.

I'm handling any additional camera related events I setup myself on closing.

Are you suggesting that for the given GenericRenderWindow instance I call `.setRenderwindow(null)`?

**Comment by jourdain** [comment]
Do you do something along those lines?

```
beforeDestroy() {
   this.view.setContainer(null);
   this.view.delete();
}
```

**Comment by mix3d** [comment]
I didn't see a .delete() in the documentation, but the documentation is just the source code, and it's hard to follow the module system's functions available to each source. 

I will try that.

**Comment by jourdain** [comment]
All instances in vtk.js are a `vtkObject` which have the following [methods](https://github.com/Kitware/vtk-js/blob/master/Sources/macro.js#L172).

**Comment by mix3d** [comment]
```
this.genericRenderWindow.setContainer(null);
this.genericRenderWindow.delete();
```
Did not work in the `beforeDestroy()` function

**Comment by jourdain** [comment]
Indeed I see where the problem lies. 
For now can you call

```
genericRW.setContainer(null);
genericRW.getOpenGLRenderWindow().delete();
genericRW.delete();
```

**Comment by mix3d** [comment]
I would recommend putting in something to the generic window to handle that.

**Comment by jourdain** [comment]
Yes the delete call should be fixed.

**Comment by mix3d** [comment]
Confirmed that deleting the openGLRenderWindow fixed the issue.

How would you update the delete function to cascade delete internal dependencies?

**Comment by jourdain** [comment]
inside the generic definition the following might work

```
publicAPI.delete = macro.chain(publicAPI.setContainer, model.openGLRenderWindow.delete, publicAPI.delete);
```

**Comment by gaston-perret** [comment]
Hello. I am having trouble with deleting OpenGL render contexts. We are working on a Vue multipage application and currently at a given page we can have a maximum of 13 active render windows. When we navigate the app, some render windows should be mounted and some destroyed, but the maximum number of active render windows shouldn´t be surpassing 13. The problem is something is not being destroyed and eventually we reach the Chrome limit (16 webgl contexts). We get the following error:
_WARNING: Too many active WebGL contexts. Oldest context will be lost._

We are now doing this, I understand that this should tell the Chrome garbage collector to delete them. Is that correct?:

`beforeDestroy () {
    this.fullscreenRenderWindow.setContainer(null)
    this.fullscreenRenderWindow.get().renderWindow.getViews()[0].delete()
    this.fullscreenRenderWindow.getOpenGLRenderWindow().delete()
    this.fullscreenRenderWindow.delete()`

Is there any other thing that could be preventing the garbage collector to clean the contexts?
Thank you for your help as usual.

**Comment by jourdain** [comment]
[This is what we do](https://github.com/Kitware/vue-vtk-js/blob/master/src/core/VtkView/script.js#L318-L357) inside our Vue component for cleanup. In your code, you are doing way more. I'm not sure why it is not working. I've noticed when working with VTK WebAssembly I could run into similar issue just because the GC was not going through fast enough. And even if you did your best to cleanup behind you, you are basically not allow to re-create new context right away. The GC need to have run and you have no control over it. 

Just make sure the canvas assigned for that view is deleted (not in DOM and not in class as ref)

**Comment by gaston-perret** [comment]
> 
> 
> [This is what we do](https://github.com/Kitware/vue-vtk-js/blob/master/src/core/VtkView/script.js#L318-L357) inside our Vue component for cleanup. In your code, you are doing way more. I'm not sure why it is not working. I've noticed when working with VTK WebAssembly I could run into similar issue just because the GC was not going through fast enough. And even if you did your best to cleanup behind you, you are basically not allow to re-create new context right away. The GC need to have run and you have no control over it.
> 
> Just make sure the canvas assigned for that view is deleted (not in DOM and not in class as ref)

Thanks for the reply. I noticed that you are using the beforeUnmount hook, and I am using beforeDestroy because in Vue 2 we cant use that hook. Could that make a difference?
Also, how could I check the canvas is deleted? I don't think I can delete it manually, its the GC that does it. All I can do is unbind everything to that canvas so the GC deletes it right?

**Comment by jourdain** [comment]
I guess I should use `beforeDestroy` too... I'm not sure if you are missing anything specific. Can you try to do a basic example reproducing that issue? Maybe using `vue-vtk-js` and see if our version actually solve the pb or not. 

What I'm worried about is that it could be a timing thing with the GC and the properly thing might be to make sure you work around that. But having such example / test-bed that could be useful for figuring out a solution. 

Otherwise depending on what you do with your views you can do something like [that](https://kitware.github.io/vtk-js/examples/ManyRenderers.html).

**Comment by gaston-perret** [comment]
> 
> 
> I guess I should use `beforeDestroy` too... I'm not sure if you are missing anything specific. Can you try to do a basic example reproducing that issue? Maybe using `vue-vtk-js` and see if our version actually solve the pb or not.
> 
> What I'm worried about is that it could be a timing thing with the GC and the properly thing might be to make sure you work around that. But having such example / test-bed that could be useful for figuring out a solution.
> 
> Otherwise depending on what you do with your views you can do something like [that](https://kitware.github.io/vtk-js/examples/ManyRenderers.html).

I think too that is a timing thing with the GC. When we load high resolution VTI (200mb) we get the error much more frequently than when we load a low resolution VTI (15mb), so probably it takes more time for the GC to free the resources.
We are now looking into refactor the app to do something like this https://kitware.github.io/vtk-js/examples/ManyRenderers.html
So we would have just 1 render window but many renderers (1 for each container/view).
Is there any downside with this? Some performance lose that we should consider? As I said we deal with big size VTI and we also have many actors and widgets in our views.

Another concern is that we have different user actions and shortcuts for each view, for example the mouse wheel does different actions in one view than the other. This was not a problem because until now we have different interactors for each view.
After this change we would have 1 interactor shared by all views.

**Comment by jourdain** [comment]
The interactor part will indeed be a tricky one.

In term of speed, the slowest rendering part will indeed dictate the rendering speed of each of the easy one. That could be fine, or it could be an issue... I don't know your app enough to judge. 

Another approach could be to do a blend of the 2 techniques. Share renderers inside a render window for specific rendering. And still having separate RW for some pieces, but just less than your original numbers. That way you don't mix and match interactor styles or heavy rendering cost? Then you rely on HTML/CSS to position things around and over each other.

**Comment by shaobeichen** [comment]
If I use vtk.js in Vue, I will encounter a crash problem due to high memory during rendering. However, in reality, if I use vtk.js+HTML separately without Vue, the same code logic will not be affected. When I use it in Vue, I placed renderWindow on this side of Vue, so I think it may be caused by this operation.

**Comment by floryst** [comment]
Hi @shaobeichen, please open a new issue with more details and code if possible.
